### PR TITLE
[Remote Store] Sync segments in refresh listener on refresh after commit

### LINKED ISSUE #10821: [Remote Store] Sync segments in refresh listener on refresh after commit
**Is your feature request related to a problem? Please describe.**
During replica to primary promotion, there is a case where the `runAfterRefreshExactlyOnce` is executed once while the syncSegments is not. The reason for this is that whenever the primary term changes, we are able to detect it by primary term check as below.
https://github.com/opensearch-project/OpenSearch/blob/14d4a6389bd5a7612c14e2e6fcb5a39822af9ee0/server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java#L165
However, the same check would not evaluate to true when the same condition is evaluated in syncSegments. This can lead to a state where the local refresh time and seq no gets updated, but the remote refresh time and seq no dont. It leads to continuous lag. Please note that this condition is mainly concerning the case when didRefresh is false and remote directory is non empty.

**Describe the solution you'd like**
In should sync method, we also check if this is a refresh after commit.

### PR DESCRIPTION
**Pull request by ashking94**
### Description
With this PR, we are doing following changes -
1. On replica promotion, we update the segments tracker with remote uploaded files on `remoteDirectory.init()` method invocation in `runAfterRefreshExactlyOnce(..)`. This ensures that we do not see spikes in refresh lag bytes on failovers.
2. There is a case during failover where the  `runAfterRefreshExactlyOnce` gets executed, but the `syncSegments` does not. This happens due to the `this.primaryTerm != indexShard.getOperationPrimaryTerm()` evaluating true for the first time, but evaluating to false on the second time.

### Related Issues
Resolves #10821, #10831

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ashking94** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/28720/
> * **CommitID:** [71e4371](https://github.com/opensearch-project/OpenSearch/commit/71e4371abd468c12e8cc5a7bcd8004ee48aaf856)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test - https://github.com/opensearch-project/OpenSearch/issues/2775

File: server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java [hunk 1]
```diff
@@ -164,16 +160,12 @@ private boolean shouldSync(boolean didRefresh) {
         // is important to upload the zero state segments so that the restore does not break.
         return this.primaryTerm != indexShard.getOperationPrimaryTerm()
             || didRefresh
-            || remoteDirectory.getSegmentsUploadedToRemoteStore().isEmpty();
+            || remoteDirectory.getSegmentsUploadedToRemoteStore().isEmpty()
+            || isRefreshAfterCommitSafe();
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:164]: why do we need this check ?

Comment (ashking94) [reply server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:164]: Added comments about this in the code itself.

Putting the same below - 
When the shouldSync is called the first time, then 1st condition on primary term is true. But after that we update the primary term and the same condition would not evaluate to true again in syncSegments. Below check ensures that if there is commit, then that gets picked up by both 1st and 2nd shouldSync call.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreStatsIT.java [hunk 2]
```diff
@@ -581,6 +601,60 @@ public void testNonZeroPrimaryStatsOnNewlyCreatedIndexWithZeroDocs() throws Exce
...
+        List<RemoteStoreStats> matches = Arrays.stream(response.getRemoteStoreStats())
+            .filter(stat -> indexShardId.equals(stat.getSegmentStats().shardId.toString()))
+            .collect(Collectors.toList());
+        assertEquals(1, matches.size());
+        RemoteSegmentTransferTracker.Stats segmentStats = matches.get(0).getSegmentStats();
+        assertEquals(0, segmentStats.refreshTimeLagMs);
+
+        networkDisruption.stopDisrupting();
+        // assertBusy(() -> assertEquals(ClusterHealthStatus.GREEN, ensureYellow(TimeValue.timeValueSeconds(1), INDEX_NAME)), 120,
+        // TimeUnit.SECONDS);
```
Comment (ashking94) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreStatsIT.java:651]: removing this

File: test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java [hunk 3]
```diff
@@ -950,6 +950,13 @@ public ClusterHealthStatus ensureYellow(String... indices) {
         return ensureColor(ClusterHealthStatus.YELLOW, TimeValue.timeValueSeconds(30), false, indices);
     }
 
+    /**
+     * Ensures the cluster has a yellow state via the cluster health API.
+     */
+    public ClusterHealthStatus ensureYellow(TimeValue timeout, String... indices) {
+        return ensureColor(ClusterHealthStatus.YELLOW, timeout, false, indices);
+    }
+
```
Comment (ashking94) [inline test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java:959]: removing this

**Review by sachinpkale — APPROVED** [review]

**Comment by ashking94** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** ABORTED ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/28772/
> * **CommitID:** [260f783](https://github.com/opensearch-project/OpenSearch/commit/260f783deebb465e2c51506deb9a3280bc818e68)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

This is a build failure before the rebase of BWC PR that was causing this issue.

**Comment by ashking94** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** UNSTABLE ❕
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.search.SearchWeightedRoutingIT.testMultiGetWithNetworkDisruption_FailOpenEnabled
> ```
> 
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/28780/
> * **CommitID:** [17b9dcd](https://github.com/opensearch-project/OpenSearch/commit/17b9dcddf60162077c53ae58ebfc9667f35d794f)
>   Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Flaky test - https://github.com/opensearch-project/OpenSearch/issues/10755.

**Pull request merged**